filedescriptor/src/unix.rs

```diff
@@ -146,6 +166,43 @@ impl OwnedHandle {
         }
     }
 
+    #[inline]
+    pub(crate) unsafe fn dup2_impl<F: AsRawFileDescriptor>(
+        fd: &F,
+        dest_fd: RawFd,
+    ) -> anyhow::Result<Self> {
+        let fd = fd.as_raw_file_descriptor();
+
+        #[cfg(not(target_os = "linux"))]
+        return Self::non_atomic_dup2(fd, dest_fd);
+
+        #[cfg(target_os = "linux")]
+        {
+            let duped = unsafe { libc::dup3(fd, dest_fd, libc::O_CLOEXEC) };
+
+            if duped == -1 {
+                let err = std::io::Error::last_os_error();
+                if let Some(libc::EINVAL) = err.raw_os_error() {
+                    // We may be running on eg: WSL or an old kernel that
+                    // doesn't support O_CLOEXEC; fall back.
+                    return Self::non_atomic_dup2(fd, dest_fd);
+                } else {
+                    bail!(
+                        "dup2 of fd {} and dest_fd {} failed: {:?}",
+                        fd,
+                        dest_fd,
+                        err
+                    )
+                }
+            } else {
+                return Ok(OwnedHandle {
+                    handle: duped,
+                    handle_type: Self::probe_handle_type(duped),
```

Comment:
I think this can match the code in `non_atomic_dup2` above:

```suggestion
                    handle_type: (),
```

Reply:
Done